### PR TITLE
[V2] test: add unit tests to pv.go

### PR DESCRIPTION
**Pull request by hccheng72**
**What type of PR is this?**
/kind test

**What this PR does / why we need it**:
Add unit tests to `pkg/controller/pv.go`

**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: pkg/controller/pv.go [hunk 1]
```diff
@@ -114,7 +113,7 @@ func (r *ReconcilePV) Reconcile(ctx context.Context, request reconcile.Request)
 
 		}
 		// AzVolume does exist and needs to be deleted
-		// add annotation to mark AzVolumeAttachment cleanup
+		// add annotation to mark AzVolume cleanup
```
Comment (sunpa93) [inline pkg/controller/pv.go:116]: This annotation is used for `AzVolumeAttachment` clean up when `AzVolume` for pre-provisioned volume is marked for deletion. So let's leave it as AzVolumeAttachment.

File: pkg/controller/pv_test.go [hunk 2]
```diff
@@ -81,6 +83,119 @@ func TestPVControllerReconcile(t *testing.T) {
 				require.NoError(t, waitErr)
 			},
 		},
+		{
+			description: "[Success] Should not requeue when PV isn't exist.",
```
Comment (sunpa93) [inline pkg/controller/pv_test.go:87]: when PV _doesn't_ exist

File: pkg/controller/pv_test.go [hunk 3]
```diff
@@ -81,6 +83,119 @@ func TestPVControllerReconcile(t *testing.T) {
...
+
+				return controller
+			},
+			verifyFunc: func(t *testing.T, controller *ReconcilePV, result reconcile.Result, err error) {
+				require.NoError(t, err)
+				require.False(t, result.Requeue)
+			},
+		},
+		{
+			description: "[Success] Should not requeue when PV is deleted and AzVolume doesn't exist.",
```
Comment (sunpa93) [inline pkg/controller/pv_test.go:105]: when PV is _marked for deletion_

File: pkg/controller/pv_test.go [hunk 4]
```diff
@@ -81,6 +83,119 @@ func TestPVControllerReconcile(t *testing.T) {
...
+
+				return controller
+			},
+			verifyFunc: func(t *testing.T, controller *ReconcilePV, result reconcile.Result, err error) {
+				require.NoError(t, err)
+				require.False(t, result.Requeue)
+			},
+		},
+		{
+			description: "[Success] Should delete AzVolume when PV is deleted but AzVolume exists.",
```
Comment (sunpa93) [inline pkg/controller/pv_test.go:127]: PV _is marked for deletion_

File: pkg/controller/pv_test.go [hunk 5]
```diff
@@ -81,6 +83,119 @@ func TestPVControllerReconcile(t *testing.T) {
...
+			},
+			verifyFunc: func(t *testing.T, controller *ReconcilePV, result reconcile.Result, err error) {
+				require.NoError(t, err)
+				require.False(t, result.Requeue)
+				_, err = controller.azClient.DiskV1beta2().AzVolumes(controller.objectNamespace).Get(context.Background(), testPersistentVolume0Name, metav1.GetOptions{})
+				require.True(t, errors.IsNotFound(err))
+			},
+		},
+		{
+			description: "[Success] Should create AzVolume when PV exists but AzVolume isn't found.",
```
Comment (sunpa93) [inline pkg/controller/pv_test.go:154]: maybe add a short parenthesis `(for pre-provisioned volume)`?

File: pkg/controller/pv_test.go [hunk 6]
```diff
@@ -81,6 +83,119 @@ func TestPVControllerReconcile(t *testing.T) {
...
+				require.Equal(t, "true", azv.Status.Annotations[consts.PreProvisionedVolumeAnnotation])
+			},
+		},
+		{
+			description: "[Success] Should ignore non-csi volumes",
+			request:     testPersistentVolume0Request,
+			setupFunc: func(t *testing.T, mockCtl *gomock.Controller) *ReconcilePV {
+				pv := testPersistentVolume0.DeepCopy()
+				pv.Status.Phase = corev1.VolumeBound
+				pv.Spec.CSI.Driver = "non-csi-driver"
```
Comment (sunpa93) [inline pkg/controller/pv_test.go:183]: set `pv.Spec.CSI = nil` given that the test case calls for "non-csi" volumes?

File: pkg/controller/pv_test.go [hunk 7]
```diff
@@ -81,6 +83,119 @@ func TestPVControllerReconcile(t *testing.T) {
...
+				mockClients(controller.cachedClient.(*mockclient.MockClient), controller.azClient, controller.kubeClient)
+
+				return controller
+			},
+			verifyFunc: func(t *testing.T, controller *ReconcilePV, result reconcile.Result, err error) {
+				require.NoError(t, err)
+				require.False(t, result.Requeue)
+				azv, err := controller.azClient.DiskV1beta2().AzVolumes(controller.objectNamespace).Get(context.Background(), testPersistentVolume0Name, metav1.GetOptions{})
+				require.NoError(t, err)
+				require.Equal(t, "true", azv.Status.Annotations[consts.PreProvisionedVolumeAnnotation])
```
Comment (sunpa93) [inline pkg/controller/pv_test.go:174]: Consider using `require.Contains(t, azv.Status.Annotations, consts.PreProvisionedVolumeAnnotation)`

File: pkg/controller/pv_test.go [hunk 8]
```diff
@@ -94,3 +209,60 @@ func TestPVControllerReconcile(t *testing.T) {
...
+	tests := []struct {
+		description string
+		setupFunc   func(*testing.T, *gomock.Controller) *ReconcilePV
+		verifyFunc  func(*testing.T, *ReconcilePV, error)
+	}{
+		{
+			description: "[Success] Should map azVolumeName, pvName, pvClaimName.",
+			setupFunc: func(t *testing.T, mockCtl *gomock.Controller) *ReconcilePV {
+				pv := testPersistentVolume0.DeepCopy()
+				pv.Status.Phase = corev1.VolumeReleased
```
Comment (sunpa93) [inline pkg/controller/pv_test.go:223]: You can remove this line :)

File: pkg/controller/pv_test.go [hunk 9]
```diff
@@ -94,3 +209,60 @@ func TestPVControllerReconcile(t *testing.T) {
...
+
+				controller := newTestPVController(
+					mockCtl,
+					testNamespace,
+					pv)
+
+				mockClients(controller.cachedClient.(*mockclient.MockClient), controller.azClient, controller.kubeClient)
+
+				pvClaimName := getQualifiedName(testNamespace, testPersistentVolumeClaim0Name)
+				controller.pvToVolumeMap.Delete(testPersistentVolume0Name)
```
Comment (sunpa93) [inline pkg/controller/pv_test.go:233]: Consider using `deletePV` and `deleteVolumeAndClaim`

File: pkg/controller/pv.go (hunk 1 above)
Comment (hccheng72) [reply pkg/controller/pv.go:116]: Done!

File: pkg/controller/pv_test.go (hunk 6 above)
Comment (hccheng72) [reply pkg/controller/pv_test.go:183]: Yeah, make sense.

Comment (hccheng72) [reply pkg/controller/pv_test.go:183]: Also edit `initState` function in `common_test.go` accordingly to prevent nil pointer dereference.

**Review by sunpa93 — APPROVED** [review]
/lgtm

**Pull request merged**